### PR TITLE
adds embedded authentication method and examples

### PR DESCRIPTION
**Pull request by anathan**
This adds support for an "embedded" authentication scheme in where devices can provide an internal serial number, signed by a shared secret, to authenticate to receive Polaris corrections

### CODE REVIEW

File: examples/simple_polaris_client.cc [hunk 1]
```diff
@@ -45,8 +44,12 @@ int main(int argc, char* argv[]) {
     return 1;
   }
 
+  auto connection_settings = point_one::polaris::DEFAULT_CONNECTION_SETTINGS;
+  connection_settings.api_host = "api.p1beta.com";
+  connection_settings.host = "polaris.p1beta.com";
```
Comment (adamshapiro0) [inline examples/simple_polaris_client.cc:49]: Should this be prod?

Comment (adamshapiro0) [reply examples/simple_polaris_client.cc:49]: Or better yet, use the default values?

File: include/point_one/polaris_asio_embedded_client.h [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+#include <memory>
+#include <mutex>
+#include <ostream>
+#include <string>
+#include <vector>
+
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+
+#include "glog/logging.h"
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:24]: Use <> for 3rd party includes

File: include/point_one/polaris_asio_embedded_client.h [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+    } catch (boost::system::system_error &e) {
+      LOG(WARNING) << "Error connecting to Polaris at tcp://"
+                   << connection_settings_.host << ":"
+                   << connection_settings_.port;
+      ScheduleReconnect();
+      return;
+    }
+
+    socket_.async_connect(endpoint_,
+                          boost::bind(&PolarisAsioEmbeddedClient::HandleConnect, this,
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:98]: 80 chars

File: include/point_one/polaris_asio_embedded_client.h [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+  void SetPolarisBytesReceived(
+      std::function<void(uint8_t *, uint16_t)> callback) {
+    polaris_bytes_received_callback_ = callback;
+  }
+
+  // Send a position to Polaris, potentially updating the resulting stream.
+  // Store the position for resending until position is updated.  ECEF
+  // variant.
+  void SetPositionECEF(double x_meters, double y_meters, double z_meters) {
+    io_service_.post(boost::bind(&PolarisAsioEmbeddedClient::HandleSetPosition, this,
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:113]: 80 chars

File: include/point_one/polaris_asio_embedded_client.h [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+                                 x_meters, y_meters, z_meters));
+  }
+
+
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:117]: Remove extra blank line

File: include/point_one/polaris_asio_embedded_client.h [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+      boost::asio::ip::tcp::socket socket(io_service_);
+      boost::asio::connect(socket, endpoint_iterator);
+
+      // Form the request. We specify the "Connection: close" header so that the
+      // server will close the socket after transmitting the response.
+      boost::asio::streambuf request;
+      std::ostream request_stream(&request);
+
+
+
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:141]: Remove exrta lines

File: include/point_one/polaris_asio_embedded_client.h [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+        LOG(ERROR) << "Invalid HTTP response to challenge request.";
+        return "";
+      }
+
+      switch(status_code) {
+        case 200:
+          VLOG(1) << "Received polaris challenge.";
+          break;
+        default:
+          LOG(ERROR) << "Challenge resquest returned with status code "
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:178]: resquest

File: include/point_one/polaris_asio_embedded_client.h [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+      if (response.size() > 0) {
+        reply_body << &response;
+      }
+      boost::system::error_code error;
+      while (boost::asio::read(socket, response,
+                               boost::asio::transfer_at_least(1), error)) {
+        reply_body << &response;
+      }
+      if (error != boost::asio::error::eof) {
+        LOG(ERROR) << "Error occured while receiving challenge: "
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:202]: occurred

File: include/point_one/polaris_asio_embedded_client.h [hunk 9]
```diff
@@ -0,0 +1,657 @@
...
+
+      std::string challenge = "";
+      try {
+        // Parse json.
+        boost::property_tree::ptree pt;
+        boost::property_tree::read_json(reply_body, pt);
+        challenge = pt.get<std::string>("challenge");
+        VLOG(3) << "Challenge is : " << challenge;
+      } catch (std::exception &e) {
+        LOG(ERROR) << "Exception in parsing challenge request response: " << e.what();
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:215]: 80 chars

File: include/point_one/polaris_asio_embedded_client.h [hunk 10]
```diff
@@ -0,0 +1,657 @@
...
+    }
+  }
+
+  // Sends authentication to the tcp server.  Result is bound to async callback
+  // HandleAuthTokenWrite.
+  void SendAuth() {
+    VLOG(1) << "Authenticating with service using auth token.";
+    if (api_token_.access_token.empty()) {
+      std::string challenge  = RequestChallenge();
+      LOG(INFO) << " GOT CHALLENGE " << challenge;
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:398]: This seems like it should be VLOG? Or at least, a cleaner formatted message (and with no leading space) if this is something a user needs to see.

File: include/point_one/polaris_asio_embedded_client.h [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+      uint64_t time = std::time(0);
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:399]: `nullptr`, not 0

File: include/point_one/polaris_asio_embedded_client.h [hunk 12]
```diff
@@ -0,0 +1,657 @@
...
+    } catch (std::exception &e) {
+      LOG(ERROR) << "Exception: " << e.what();
+    }
+    return "";
+  }
+
+  std::string BuildChallengeSignature(std::string challenge, uint64_t time) {
+    // concatenate the string and calculate the message digest
+
+    // challenge| timestamp | company_id | serial_number
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:229]: missing space after challenge

File: include/point_one/polaris_asio_embedded_client.h [hunk 13]
```diff
@@ -0,0 +1,657 @@
...
+    std::stringstream ss;
+    ss << challenge << std::to_string(time) << company_id_ << unique_id_;
+
+    unsigned int HMAC_SHA512_HASH_LEN = 64;
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:233]: `static constexpr`

File: include/point_one/polaris_asio_embedded_client.h [hunk 14]
```diff
@@ -0,0 +1,657 @@
...
+    uint8_t buf[HMAC_SHA512_HASH_LEN];
+    std::string message = ss.str();
+    HMAC_CTX hmac_ctx;
+    HMAC_CTX_init(&hmac_ctx);
+    HMAC_Init_ex(&hmac_ctx, (void *)signing_secret_.data(),
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:238]: Do you actually need to cast to void*? That's surprising - that's usually the reason you use void* in prototypes like this.

File: include/point_one/polaris_asio_embedded_client.h [hunk 15]
```diff
@@ -0,0 +1,657 @@
...
+                 signing_secret_.length(), EVP_sha512(), NULL);
+    HMAC_Update(&hmac_ctx, (uint8_t *)message.data(), message.length());
+    HMAC_Final(&hmac_ctx, buf, &HMAC_SHA512_HASH_LEN);
+    unsigned char encodedData[128] = {};  // this is beyond safe, the max length
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:242]: Maybe move the blank line below to above this line to separate HMAC stuff from EVP stuff

File: include/point_one/polaris_asio_embedded_client.h [hunk 16]
```diff
@@ -0,0 +1,657 @@
...
+                                          // of this is 1.25 * the HASH LEN
+
+    EVP_EncodeBlock(encodedData, buf, HMAC_SHA512_HASH_LEN);
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:245]: All of these functions have return codes and error conditions. Should we be checking them and handling them somehow?

Also, this function returns the number of bytes written to the output. Consider storing that and passing to the string constructor.

File: include/point_one/polaris_asio_embedded_client.h (hunk 15 above)
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:242]: Naming style: encoded_data

File: include/point_one/polaris_asio_embedded_client.h [hunk 17]
```diff
@@ -0,0 +1,657 @@
...
+    std::string output((char *)encodedData);
+    VLOG(2) << "Challenge Response Signature :" << std::endl << output;
+    return output;
+  }
+
+  bool GetAuthToken(std::string challenge, uint64_t time, std::string signature) {
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:251]: 80 chars

File: include/point_one/polaris_asio_embedded_client.h [hunk 18]
```diff
@@ -0,0 +1,657 @@
...
+      // Try each endpoint until we successfully establish a connection.
+      boost::asio::ip::tcp::socket socket(io_service_);
+      boost::asio::connect(socket, endpoint_iterator);
+
+      // Form the request. We specify the "Connection: close" header so that the
+      // server will close the socket after transmitting the response.
+      boost::asio::streambuf request;
+      std::ostream request_stream(&request);
+
+
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:269]: Extra line

File: include/point_one/polaris_asio_embedded_client.h [hunk 19]
```diff
@@ -0,0 +1,657 @@
...
+      std::stringstream post_body;
+      post_body << "{\"challenge\": \"" << challenge << "\", "
+                << "\"timestamp\":  \"" << std::to_string(time) << "\", "
+                << "\"company_id\": \"" << company_id_ << "\", "
+                << "\"unique_id\": \"" << unique_id_ << "\", "
+                << "\"signature\": \"" << signature << "\""
+                << "}\r\n";
+      std::string post_body_str = post_body.str();
+      LOG(INFO) << " JSON " << std::endl << post_body_str;
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:278]: This should be a low level VLOG, and if you keep it at all it really needs more context.

File: include/point_one/polaris_asio_embedded_client.h [hunk 20]
```diff
@@ -0,0 +1,657 @@
...
+      request_stream << "Content-Type: application/json\r\n";
+      request_stream << "Content-Length: " << post_body_str.size()
+                     << "\r\n";
+      request_stream << "Accept: */*\r\n";
+      request_stream << "Connection: close\r\n";
+      request_stream << "Cache-Control: no-cache\r\n";
+      request_stream << "\r\n";
+      request_stream << post_body_str;
+
+
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:293]: Extra line

File: include/point_one/polaris_asio_embedded_client.h [hunk 21]
```diff
@@ -0,0 +1,657 @@
...
+        VLOG(3) << "Access token: " << api_token_.access_token;
+        VLOG(3) << "Expires in: " << std::fixed << api_token_.expires_in;
+        VLOG(3) << "Issued at: " << std::fixed << api_token_.issued_at;
+
+      } catch (std::exception &e) {
+        LOG(ERROR) << "Exception in parsing auth token response: " << e.what();
+        return false;
+      }
+
+      return true;
```
Comment (adamshapiro0) [inline include/point_one/polaris_asio_embedded_client.h:370]: Move this into the try

File: include/point_one/polaris_asio_embedded_client.h (hunk 14 above)
Comment (anathan) [reply include/point_one/polaris_asio_embedded_client.h:238]: that was a leftover, good catch.

**Comment by anathan** [comment]
Definitely agree on the duplication. Let's put that in the next revision to break out the client into a few smaller, modular pieces.

**Pull request merged**